zephyr/lib/regions_mm.c

```diff
@@ -1,53 +1,428 @@
+		return -ENOTEMPTY;
+
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++) {
+		if (heap->physical_blocks_allocators[i] != 0) {
+			rfree(heap->physical_blocks_allocators[i]->bitmap->bundles);
+			rfree(heap->physical_blocks_allocators[i]->bitmap);
+			rfree(heap->physical_blocks_allocators[i]);
+		}
+	}
+	rfree(heap);
+	return 0;
+}
+
+/**
+ * @brief Free ptr allocated on given heap
+ *
+ * Free the ptr allocation. After free action is complete
+ * check if any physical memory block can be freed up as
+ * result.
+ * @param heap Ptr to the heap that is supposed to be freed
+ * @param ptr pointer that should be freed
+ *
+ * @retval 0 on success;
+ * @retval -ENOTEMPTY on heap having active allocations.
+ */
+int vmh_free(const struct virtual_memory_heap *heap, void *ptr)
+{
+	struct list_item *iterator;
+	struct active_allocation *allocation = NULL;
+	int retval;
+
+	list_for_item(iterator, &heap->active_allocations)
+	{
+		if (ptr == (uint8_t *)CONTAINER_OF(iterator,
+			struct active_allocation, node)->allocation_ptr) {
+			allocation = (struct active_allocation *)
+				CONTAINER_OF(iterator,
+					struct active_allocation, node)->allocation_ptr;
+			break;
+		}
+	}
+
+	if (allocation == NULL)
+		return -EEXIST;
+
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++) {
+		if (is_ptr_in_range(ptr,
+			heap->physical_blocks_allocators[i]->buffer,
+			heap->physical_blocks_allocators[i]->num_blocks *
+			(heap->physical_blocks_allocators[i]->blk_sz_shift << 1))) {
+			if (heap->span)	{
+				retval = sys_mem_blocks_free_contiguous(
+					heap->physical_blocks_allocators[i], ptr,
+					allocation->allocation_size);
+			} else {
+				retval = sys_mem_blocks_free(
+					heap->physical_blocks_allocators[i],
+					allocation->allocation_size, (void **)&ptr);
+			}
 			break;
-		case MEM_REG_ATTR_OPPORTUNISTIC_MEMORY:
-			vm_heaps[i].memory_caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP;
+		}
+	}
+
+	list_item_del(&allocation->node);
+	rfree(CONTAINER_OF(&allocation->node, struct active_allocation, node));
+
+	/* Scan for changes in physical pages mapping and
+	 * apply unmap if possible
+	 */
+	struct list_item *iterator_phys;
+	struct list_item *tmp;
+	struct list_item *iterator_activ;
+
+	list_for_item_safe(iterator_phys, tmp, &physical_mappings)
+	{
+		bool can_free = true;
+
+		list_for_item(iterator_activ, &heap->active_allocations)
+		{
+			struct active_physical_mapping *physical_maping = CONTAINER_OF(
+				iterator_phys, struct active_physical_mapping, node);
+			struct active_allocation *active_alloc = CONTAINER_OF(
+				iterator_activ, struct active_allocation, node);
+
+			if (is_range_in_range(
+				active_alloc->allocation_ptr,
+				active_alloc->allocation_size,
+				physical_maping->mapping_ptr,
+				physical_maping->mapping_size)) {
+
+				can_free = false;
+				break;
+			}
+		}
+		if (can_free) {
+			sys_mm_drv_unmap_region(CONTAINER_OF(iterator_phys,
+				struct active_physical_mapping, node)->mapping_ptr,
```

Comment:
Assign `CONTAINER_OF(iterator_phys, struct active_physical_mapping, node)` to a variable.